src/conn/conn_oligarch.c

```diff
@@ -8,192 +8,97 @@
-
-        /* Parse out the key and new checkpoint config */
-        last_sep = 0;
-        for (new_sz = 4095; new_sz >= 0; new_sz--) {
-            if (buf[new_sz] == '|') {
-                last_sep = new_sz;
-                break;
-            }
         }
+    if (metadata_value == NULL)
+        WT_ERR(EINVAL);
 
-        name_ptr = last_sep;
-        while (name_ptr != 0 && buf[name_ptr - 1] != '\n')
-            name_ptr--;
+    /* We need an internal session when modifying metadata. */
+    WT_RET(__wt_open_internal_session(conn, "checkpoint-pick-up", false, 0, 0, &internal_session));
 
-#if 0
-        /* XXX Get the checkpoint metadata from dir_store - just to check that it all works. */
-        __wt_sleep(0, 1000);
-        WT_ASSERT(session, __wt_disagg_get_meta(session, 0, 0, item) == 0);
-        WT_ASSERT(session, memcmp(item->data, buf + name_ptr, item->size) == 0);
-#endif
-
-        buf[last_sep] = '\0';
-
-        /* Open up a metadata cursor pointing at our table */
-        WT_ERR(__wt_metadata_cursor(session, &md_cursor));
-        md_cursor->set_key(md_cursor, &buf[name_ptr]);
-        WT_ERR(md_cursor->search(md_cursor));
-
-        /* Pull the value out */
-        WT_ERR(md_cursor->get_value(md_cursor, &value));
-        len = strlen(&buf[last_sep + 1]);
-        buf[last_sep + (wt_off_t)len] = '\0'; /* lop off the trailing newline */
-        len += strlen("checkpoint=");         /* -1 for trailing newline */
-
-        /* Allocate/create a new config we're going to insert */
-        WT_ERR(__wt_calloc_def(session, len, &new_md_value));
-        WT_ERR(__wt_snprintf(new_md_value, len, "checkpoint=%s", &buf[last_sep + 1]));
-        /* fprintf(stderr, "[%s] loading metadata %s\n", S2C(session)->home, new_md_value); */
-        cfg[0] = value;
-        cfg[1] = new_md_value;
-        cfg[2] = NULL;
-        WT_ERR(__wt_config_collapse(session, cfg, &cfg_ret));
-
-        /* Put our new config in */
-        WT_ERR(__wt_metadata_insert(session, &buf[name_ptr], cfg_ret));
-        WT_ERR(__wt_metadata_cursor_release(session, &md_cursor));
-        S2C(session)->oligarch_manager.update_dhandle =
-          true; /* TODO concurrency hazard, needs better design */
-        md_cursor = NULL;
+    /* Open up a metadata cursor pointing at our table */
+    WT_ERR(__wt_metadata_cursor(internal_session, &md_cursor));
+    md_cursor->set_key(md_cursor, metadata_key);
+    WT_ERR(md_cursor->search(md_cursor));
 
-        /*
-         * WiredTiger will reload the dir store's checkpoint when opening a cursor: Opening a file
-         * cursor triggers __wt_btree_open (even if the file has been opened before).
-         */
-        WT_STAT_CONN_DSRC_INCR(session, oligarch_manager_checkpoints_refreshed);
-    }
+    /* Pull the value out. */
+    WT_ERR(md_cursor->get_value(md_cursor, &current_value));
+    len = strlen(metadata_value) + strlen("checkpoint=") + 1 /* for NUL */;
```

Comment:
Doesn't it need a comma, like: `strlen(",checkpoint=")` ?

Reply:
The constructed string just starts with "checkpoint=", but I see it's confusing, so I will swap the order.